pkg/controller/master/upgrade/upgrade_controller.go

```diff
@@ -461,55 +462,64 @@ func (h *upgradeHandler) cleanup(upgrade *harvesterv1.Upgrade, cleanJobs bool) e
 			}
 			jobs, err := h.jobCache.List(plan.Namespace, set.AsSelector())
 			if err != nil {
-				return err
+				return nil, err
 			}
 			for _, job := range jobs {
 				logrus.Debugf("Deleting job %s/%s", job.Namespace, job.Name)
 				if err := h.jobClient.Delete(job.Namespace, job.Name, &metav1.DeleteOptions{}); err != nil {
-					return err
+					return nil, err
 				}
 			}
 		}
 
 		logrus.Debugf("Deleting plan %s/%s", plan.Namespace, plan.Name)
 		if err := h.planClient.Delete(plan.Namespace, plan.Name, &metav1.DeleteOptions{}); err != nil {
-			return err
+			return nil, err
 		}
 	}
 
 	// restore Longhorn replica-replenishment-wait-interval and
 	// auto-cleanup-system-generated-snapshot settings (multi-node cluster only)
+	toUpdate := upgrade.DeepCopy()
 	if upgrade.Status.SingleNode == "" {
-		if err := h.loadReplicaReplenishmentFromUpgradeAnnotation(upgrade); err != nil {
-			return err
-		}
-		if err := h.loadAutoCleanupSystemGeneratedSnapshotFromUpgradeAnnotation(upgrade); err != nil {
-			return err
+		_, exists := upgrade.Annotations[longhornSettingsRestoredAnnotation]
+		if !exists {
+			if err := h.loadReplicaReplenishmentFromUpgradeAnnotation(upgrade); err != nil {
+				return nil, err
+			}
+			if err := h.loadAutoCleanupSystemGeneratedSnapshotFromUpgradeAnnotation(upgrade); err != nil {
+				return nil, err
+			}
+			toUpdate.Annotations[longhornSettingsRestoredAnnotation] = strconv.FormatBool(true)
```

Comment:
the `toUpdate` will be write back on L522, but there are other possible failures on L502, L512, L519,  in those cases, LH setting is amended, but upgrade CR itself does not book it;  in next retry, it will try to write to LH again if LH's value is different with current (e.g., set by user)

two options:
(1) Write to upgrade CR on L494, and let next OnChange to run following upgradeLog, managedchart
(2) Change L502, L512, L519, which will write back anyway even when error happens, but sill return the err, to let reconciller retry